Python/ceval.c

```diff
@@ -2726,31 +2726,6 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
-        TARGET(POP_EXCEPT_AND_RERAISE) {
-            PyObject *lasti = PEEK(2);
-            if (PyLong_Check(lasti)) {
-                frame->f_lasti = PyLong_AsLong(lasti);
-                assert(!_PyErr_Occurred(tstate));
-            }
-            else {
-                _PyErr_SetString(tstate, PyExc_SystemError, "lasti is not an int");
-                goto error;
-            }
-            PyObject *value = POP();
-            assert(value);
-            assert(PyExceptionInstance_Check(value));
-            PyObject *type = Py_NewRef(PyExceptionInstance_Class(value));
-            PyObject *traceback = PyException_GetTraceback(value);
-            Py_DECREF(POP()); /* lasti */
-            _PyErr_Restore(tstate, type, value, traceback);
-
-            _PyErr_StackItem *exc_info = tstate->exc_info;
-            value = exc_info->exc_value;
-            exc_info->exc_value = POP();
-            Py_XDECREF(value);
-            goto exception_unwind;
-        }
```

Comment:
IIUC, this PR changes the order of operations but should preserve the "before" and "after":

1. re-order the stack
2. pop (C)
3. set the "currently handling" exception to (C), replacing (X)
4. check (B)
5. set frame->f_lasti to the int from (B)
6. pop (A)
7. check (A)
8. set the current (unhandled) exception to (A)

Does that look right?

Reply:
So, in the old code we were popping off (B) (and decrefing it), but I don't see where that happens with the new code.  Did I miss something?  Regardless, I expect tests would fail if this were a problem, but everything looks okay, so (B) must get used up correctly elsewhere.

Reply:
That looks right. I think the "RERAISE 1" opcode is doing to B exactly the same as was happening here, is it not?

X is the exception we just finished handling, and we are overriding it with the previous "handled exception" (see my previous comment about nesting).

Reply:
The difference is that in `POP_EXCEPT_AND_RERAISE` (B) is popped off the stack (and decref'ed).  In `RERAISE` it isn't.

Reply:
You’re right. I think the reraise case just goes straight to unwinding the exception, which pops everything from the stack that came from this frame.

Reply:
ah, so we're okay leaving it?  or should we explicitly pop it off?

Reply:
I think it’s ok leaving it.